### DISCUSSION TITLE
Unexpected non-matching ElementGraph hashes

### DISCUSSION THREAD

**Discussion by jepdavidson** [rdkit/rdkit#6470] [Q&A]
Dear All,

I have an example of two molecules that I believe should give the same ElementGraph hashes, but they don't (well - actually they **do** using NextMove's molhash binary, but they don't using RDKit...).

The question is - is there a setting or approach that I am missing to get these two to match?

Here's the example:
![image](https://github.com/rdkit/rdkit/assets/7816101/5f6f1359-bb69-4697-8809-69eed689c43d)

```python
from rdkit import Chem, rdBase
from rdkit.Chem import rdMolHash

print(rdBase.rdkitVersion)
m1 = Chem.MolFromSmiles("C1COC(C1)C1=NC=NC=C1")
m2 = Chem.MolFromSmiles("C1CC(N=CN1)C1=CC=CO1"
hashes = [rdMolHash.MolHash(m, rdMolHash.HashFunction.ElementGraph) for m in (m1, m2)]
print(*hashes, hashes[0] == hashes[1], sep="\t")
```
> 2023.03.1
> C1COC(C2CCNCN2)C1
> C1CC(C2CCCO2)NCN1
> False

Here are the same inputs using the molhash cli (pushing the smiles strings in with newline separator, then -t flag for to only output the hash [title only option] and -e for element graph):

```bash
echo "C1COC(C1)C1=NC=NC=C1\nC1CC(N=CN1)C1=CC=CO1" | molhash -t -e -
```
> C1CC(OC1)C2CCNCN2
> C1CC(OC1)C2CCNCN2

As always, any help/advice is greatly appreciated!

Kind regards

James

**Comment by greglandrum** [comment]
Hi @jepdavidson.
It looks like this is a bug which affects both the ElementGraph and AnonymousGraph hashes (they use the same code in the background). The code is setting all of the bond types to single but isn't clearing the aromatic flag. That flag is used in the canonicalization, so you end up with different SMILES.
Clearing the aromatic flags removes the problem:
```
In [9]: for b in m1.GetBonds(): b.SetIsAromatic(False)

In [10]: for b in m2.GetBonds(): b.SetIsAromatic(False)

In [11]: len(set([rdMolHash.MolHash(m,rdMolHash.HashFunction.ElementGraph) for
    ...: m in (m1,m2)]))
Out[11]: 1
```

This is an easy fix for the 2023.03.3 release; could you please create the issue so that you get the credit for it? There's a "create issue from discussion" link here in the github UI.

**Reply by jepdavidson** [reply]
Thanks @greglandrum - I did wonder about an aromaticity cause, but did some other tests involving aromatic inputs and found no problem.  I must have been lucky getting the matching canonicalisation in those cases (or lucky not to here I guess!).

I'll create the issue now.

Kind regards

James
